{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2e8947ee-d670-4d8f-8d4f-26294b4e1bc8",
   "metadata": {
    "tags": []
   },
   "source": [
    "\n",
    "# Figure 1 Clustering and ADTs- ResCITE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c2b1393-421e-4157-824b-8d9579cb90b5",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bd4d8974-63d9-40a8-b305-665b9aaec69c",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, t\n",
      "\n",
      "\n",
      "Loading required package: data.table\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: rhdf5\n",
      "\n",
      "\n",
      "Attaching package: ‘H5weaver’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:rhdf5’:\n",
      "\n",
      "    h5dump, h5ls\n",
      "\n",
      "\n",
      "Warning message in system(\"timedatectl\", intern = TRUE):\n",
      "“running command 'timedatectl' had status 1”\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.3     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdata.table\u001b[39m::\u001b[32mbetween()\u001b[39m masks \u001b[34mdplyr\u001b[39m::between()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m       masks \u001b[34mMatrix\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m       masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdata.table\u001b[39m::\u001b[32mfirst()\u001b[39m   masks \u001b[34mdplyr\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mhour()\u001b[39m     masks \u001b[34mdata.table\u001b[39m::hour()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32misoweek()\u001b[39m  masks \u001b[34mdata.table\u001b[39m::isoweek()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m          masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdata.table\u001b[39m::\u001b[32mlast()\u001b[39m    masks \u001b[34mdplyr\u001b[39m::last()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mmday()\u001b[39m     masks \u001b[34mdata.table\u001b[39m::mday()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mminute()\u001b[39m   masks \u001b[34mdata.table\u001b[39m::minute()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mmonth()\u001b[39m    masks \u001b[34mdata.table\u001b[39m::month()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mpack()\u001b[39m         masks \u001b[34mMatrix\u001b[39m::pack()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mquarter()\u001b[39m  masks \u001b[34mdata.table\u001b[39m::quarter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond()\u001b[39m   masks \u001b[34mdata.table\u001b[39m::second()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mtranspose()\u001b[39m    masks \u001b[34mdata.table\u001b[39m::transpose()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32munpack()\u001b[39m       masks \u001b[34mMatrix\u001b[39m::unpack()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mwday()\u001b[39m     masks \u001b[34mdata.table\u001b[39m::wday()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mweek()\u001b[39m     masks \u001b[34mdata.table\u001b[39m::week()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32myday()\u001b[39m     masks \u001b[34mdata.table\u001b[39m::yday()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32myear()\u001b[39m     masks \u001b[34mdata.table\u001b[39m::year()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(dplyr)\n",
    "library(Seurat)\n",
    "library(patchwork)\n",
    "library(H5weaver)\n",
    "library(hise)\n",
    "library(tidyverse)\n",
    "library(SeuratObject)\n",
    "library(SeuratDisk)\n",
    "library(ggrepel)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c3789a1-4230-4d24-b301-2902814fab28",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Reading H5 files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3892216c-60be-4fdc-91f9-b649d7641b93",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5s <- list.files(\n",
    "    path = '/home/jupyter/CS15_WHBL/CWB_Paper/01_Final_Data/01_Data/scrublet', \n",
    "    pattern = 'filtered_feature_bc_matrix.h5$', \n",
    "    full.names = TRUE, \n",
    "    recursive = TRUE\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fb938386-9712-44c4-ab79-e210f3b52b85",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Genome matrix has multiple modalities, returning a list of matrices for this genome\n",
      "\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Genome matrix has multiple modalities, returning a list of matrices for this genome\n",
      "\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Genome matrix has multiple modalities, returning a list of matrices for this genome\n",
      "\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Genome matrix has multiple modalities, returning a list of matrices for this genome\n",
      "\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Genome matrix has multiple modalities, returning a list of matrices for this genome\n",
      "\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Genome matrix has multiple modalities, returning a list of matrices for this genome\n",
      "\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Genome matrix has multiple modalities, returning a list of matrices for this genome\n",
      "\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Genome matrix has multiple modalities, returning a list of matrices for this genome\n",
      "\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n"
     ]
    }
   ],
   "source": [
    "fully <- lapply(h5s, function(x){\n",
    "\n",
    "    mtx <- Read10X_h5(x) # reading the matrix\n",
    "    \n",
    "    so <- CreateSeuratObject(mtx$`Gene Expression`)\n",
    "    \n",
    "    so[['ADT']] <- CreateAssayObject(mtx$`Antibody Capture`)\n",
    "    \n",
    "    so$orig.ident <- strsplit(x,'/')[[1]][9]\n",
    "    so$orig.ident <- strsplit(so$orig.ident,'_')[[1]][1]\n",
    "    so$type <- substr(so$orig.ident,0,2)\n",
    "\n",
    "    return(so) \n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7727040d-7868-4866-96f2-9dc319b535ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "scrub <- list.files(\n",
    "    # Place your analysis in a folder by number experiment, or whatever might work for you\n",
    "    # It helps if this worksheet in in the same folder\n",
    "    path = '/home/jupyter/CS15_WHBL/CWB_Paper/01_Final_Data/01_Data/scrublet',\n",
    "    pattern = '.csv$', # change this to 'sample_raw_feature_bc_matrix.h5$' if you need the raw h5s\n",
    "    full.names = TRUE, \n",
    "    recursive = TRUE\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "838b1ddb-994a-4cee-b667-4704311e72d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "scrubs <- lapply(scrub,read.csv)\n",
    "\n",
    "scrubs <- scrubs %>% lapply(function(x){\n",
    "    rownames(x) <- x$X\n",
    "    x$X <- NULL\n",
    "    x$Barcodes <- rownames(x)\n",
    "    return(x)\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3659dc0d-fe5e-4a8b-9a5a-cf7acb1b26ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (i in c(1:8)){\n",
    "    \n",
    "    fully[[i]][[]]$Barcodes <- rownames(fully[[i]][[]])\n",
    "    \n",
    "    fully[[i]][[]] <- left_join(fully[[i]][[]], scrubs[[i]], by = 'Barcodes') # , by='row.names', all=TRUE\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1897dc57-3ca4-457d-a000-b5b18a59034b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fully <- Reduce(merge,fully) %>%\n",
    "    subset(subset=Predicted_Doublet=='False') %>% \n",
    "    JoinLayers()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ee02d031-6dff-4a82-80c3-9da2a6e871ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 10 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th><th scope=col>nCount_ADT</th><th scope=col>nFeature_ADT</th><th scope=col>type</th><th scope=col>Barcodes</th><th scope=col>Doublet_Score</th><th scope=col>Predicted_Doublet</th><th scope=col>Doublet_Score.x</th><th scope=col>Predicted_Doublet.x</th><th scope=col>Doublet_Score.y</th><th scope=col>Predicted_Doublet.y</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>AAACAAGCAATTGAACACTTTAGG-1</th><td>BL05111</td><td>13302</td><td>4388</td><td> 13535</td><td>127</td><td>BL</td><td>AAACAAGCAATTGAACACTTTAGG-1</td><td>0.03723764</td><td>False</td><td>0.03723764</td><td>False</td><td>0.03723764</td><td>False</td></tr>\n",
       "\t<tr><th scope=row>AAACAAGCACAAATACACTTTAGG-1</th><td>BL05111</td><td> 6375</td><td>3124</td><td> 10241</td><td>125</td><td>BL</td><td>AAACAAGCACAAATACACTTTAGG-1</td><td>0.08790072</td><td>False</td><td>0.08790072</td><td>False</td><td>0.08790072</td><td>False</td></tr>\n",
       "\t<tr><th scope=row>AAACAAGCATACTTCAACTTTAGG-1</th><td>BL05111</td><td>  625</td><td> 431</td><td>144311</td><td>130</td><td>BL</td><td>AAACAAGCATACTTCAACTTTAGG-1</td><td>0.01258178</td><td>False</td><td>0.01258178</td><td>False</td><td>0.01258178</td><td>False</td></tr>\n",
       "\t<tr><th scope=row>AAACCAATCACTATGTACTTTAGG-1</th><td>BL05111</td><td> 6118</td><td>3041</td><td> 17478</td><td>129</td><td>BL</td><td>AAACCAATCACTATGTACTTTAGG-1</td><td>0.09899889</td><td>False</td><td>0.09899889</td><td>False</td><td>0.09899889</td><td>False</td></tr>\n",
       "\t<tr><th scope=row>AAACCAATCATGAGGTACTTTAGG-1</th><td>BL05111</td><td>22733</td><td>5900</td><td> 14822</td><td>128</td><td>BL</td><td>AAACCAATCATGAGGTACTTTAGG-1</td><td>0.03723764</td><td>False</td><td>0.03723764</td><td>False</td><td>0.03723764</td><td>False</td></tr>\n",
       "\t<tr><th scope=row>AAACCAATCCGCTATAACTTTAGG-1</th><td>BL05111</td><td> 9181</td><td>3722</td><td> 12917</td><td>124</td><td>BL</td><td>AAACCAATCCGCTATAACTTTAGG-1</td><td>0.09052632</td><td>False</td><td>0.09052632</td><td>False</td><td>0.09052632</td><td>False</td></tr>\n",
       "\t<tr><th scope=row>AAACCAATCTGTGACCACTTTAGG-1</th><td>BL05111</td><td> 7964</td><td>3618</td><td> 19663</td><td>128</td><td>BL</td><td>AAACCAATCTGTGACCACTTTAGG-1</td><td>0.03614458</td><td>False</td><td>0.03614458</td><td>False</td><td>0.03614458</td><td>False</td></tr>\n",
       "\t<tr><th scope=row>AAACCAGGTGACATTCACTTTAGG-1</th><td>BL05111</td><td> 8798</td><td>3849</td><td> 10989</td><td>125</td><td>BL</td><td>AAACCAGGTGACATTCACTTTAGG-1</td><td>0.03507611</td><td>False</td><td>0.03507611</td><td>False</td><td>0.03507611</td><td>False</td></tr>\n",
       "\t<tr><th scope=row>AAACGGGCAAACTGGCACTTTAGG-1</th><td>BL05111</td><td> 8204</td><td>3595</td><td> 12555</td><td>121</td><td>BL</td><td>AAACGGGCAAACTGGCACTTTAGG-1</td><td>0.02644231</td><td>False</td><td>0.02644231</td><td>False</td><td>0.02644231</td><td>False</td></tr>\n",
       "\t<tr><th scope=row>AAACGGGCAGGATCCGACTTTAGG-1</th><td>BL05111</td><td>  840</td><td> 674</td><td> 70401</td><td>130</td><td>BL</td><td>AAACGGGCAGGATCCGACTTTAGG-1</td><td>0.02151755</td><td>False</td><td>0.02151755</td><td>False</td><td>0.02151755</td><td>False</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 13\n",
       "\\begin{tabular}{r|lllllllllllll}\n",
       "  & orig.ident & nCount\\_RNA & nFeature\\_RNA & nCount\\_ADT & nFeature\\_ADT & type & Barcodes & Doublet\\_Score & Predicted\\_Doublet & Doublet\\_Score.x & Predicted\\_Doublet.x & Doublet\\_Score.y & Predicted\\_Doublet.y\\\\\n",
       "  & <chr> & <dbl> & <int> & <dbl> & <int> & <chr> & <chr> & <dbl> & <chr> & <dbl> & <chr> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\tAAACAAGCAATTGAACACTTTAGG-1 & BL05111 & 13302 & 4388 &  13535 & 127 & BL & AAACAAGCAATTGAACACTTTAGG-1 & 0.03723764 & False & 0.03723764 & False & 0.03723764 & False\\\\\n",
       "\tAAACAAGCACAAATACACTTTAGG-1 & BL05111 &  6375 & 3124 &  10241 & 125 & BL & AAACAAGCACAAATACACTTTAGG-1 & 0.08790072 & False & 0.08790072 & False & 0.08790072 & False\\\\\n",
       "\tAAACAAGCATACTTCAACTTTAGG-1 & BL05111 &   625 &  431 & 144311 & 130 & BL & AAACAAGCATACTTCAACTTTAGG-1 & 0.01258178 & False & 0.01258178 & False & 0.01258178 & False\\\\\n",
       "\tAAACCAATCACTATGTACTTTAGG-1 & BL05111 &  6118 & 3041 &  17478 & 129 & BL & AAACCAATCACTATGTACTTTAGG-1 & 0.09899889 & False & 0.09899889 & False & 0.09899889 & False\\\\\n",
       "\tAAACCAATCATGAGGTACTTTAGG-1 & BL05111 & 22733 & 5900 &  14822 & 128 & BL & AAACCAATCATGAGGTACTTTAGG-1 & 0.03723764 & False & 0.03723764 & False & 0.03723764 & False\\\\\n",
       "\tAAACCAATCCGCTATAACTTTAGG-1 & BL05111 &  9181 & 3722 &  12917 & 124 & BL & AAACCAATCCGCTATAACTTTAGG-1 & 0.09052632 & False & 0.09052632 & False & 0.09052632 & False\\\\\n",
       "\tAAACCAATCTGTGACCACTTTAGG-1 & BL05111 &  7964 & 3618 &  19663 & 128 & BL & AAACCAATCTGTGACCACTTTAGG-1 & 0.03614458 & False & 0.03614458 & False & 0.03614458 & False\\\\\n",
       "\tAAACCAGGTGACATTCACTTTAGG-1 & BL05111 &  8798 & 3849 &  10989 & 125 & BL & AAACCAGGTGACATTCACTTTAGG-1 & 0.03507611 & False & 0.03507611 & False & 0.03507611 & False\\\\\n",
       "\tAAACGGGCAAACTGGCACTTTAGG-1 & BL05111 &  8204 & 3595 &  12555 & 121 & BL & AAACGGGCAAACTGGCACTTTAGG-1 & 0.02644231 & False & 0.02644231 & False & 0.02644231 & False\\\\\n",
       "\tAAACGGGCAGGATCCGACTTTAGG-1 & BL05111 &   840 &  674 &  70401 & 130 & BL & AAACGGGCAGGATCCGACTTTAGG-1 & 0.02151755 & False & 0.02151755 & False & 0.02151755 & False\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 13\n",
       "\n",
       "| <!--/--> | orig.ident &lt;chr&gt; | nCount_RNA &lt;dbl&gt; | nFeature_RNA &lt;int&gt; | nCount_ADT &lt;dbl&gt; | nFeature_ADT &lt;int&gt; | type &lt;chr&gt; | Barcodes &lt;chr&gt; | Doublet_Score &lt;dbl&gt; | Predicted_Doublet &lt;chr&gt; | Doublet_Score.x &lt;dbl&gt; | Predicted_Doublet.x &lt;chr&gt; | Doublet_Score.y &lt;dbl&gt; | Predicted_Doublet.y &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| AAACAAGCAATTGAACACTTTAGG-1 | BL05111 | 13302 | 4388 |  13535 | 127 | BL | AAACAAGCAATTGAACACTTTAGG-1 | 0.03723764 | False | 0.03723764 | False | 0.03723764 | False |\n",
       "| AAACAAGCACAAATACACTTTAGG-1 | BL05111 |  6375 | 3124 |  10241 | 125 | BL | AAACAAGCACAAATACACTTTAGG-1 | 0.08790072 | False | 0.08790072 | False | 0.08790072 | False |\n",
       "| AAACAAGCATACTTCAACTTTAGG-1 | BL05111 |   625 |  431 | 144311 | 130 | BL | AAACAAGCATACTTCAACTTTAGG-1 | 0.01258178 | False | 0.01258178 | False | 0.01258178 | False |\n",
       "| AAACCAATCACTATGTACTTTAGG-1 | BL05111 |  6118 | 3041 |  17478 | 129 | BL | AAACCAATCACTATGTACTTTAGG-1 | 0.09899889 | False | 0.09899889 | False | 0.09899889 | False |\n",
       "| AAACCAATCATGAGGTACTTTAGG-1 | BL05111 | 22733 | 5900 |  14822 | 128 | BL | AAACCAATCATGAGGTACTTTAGG-1 | 0.03723764 | False | 0.03723764 | False | 0.03723764 | False |\n",
       "| AAACCAATCCGCTATAACTTTAGG-1 | BL05111 |  9181 | 3722 |  12917 | 124 | BL | AAACCAATCCGCTATAACTTTAGG-1 | 0.09052632 | False | 0.09052632 | False | 0.09052632 | False |\n",
       "| AAACCAATCTGTGACCACTTTAGG-1 | BL05111 |  7964 | 3618 |  19663 | 128 | BL | AAACCAATCTGTGACCACTTTAGG-1 | 0.03614458 | False | 0.03614458 | False | 0.03614458 | False |\n",
       "| AAACCAGGTGACATTCACTTTAGG-1 | BL05111 |  8798 | 3849 |  10989 | 125 | BL | AAACCAGGTGACATTCACTTTAGG-1 | 0.03507611 | False | 0.03507611 | False | 0.03507611 | False |\n",
       "| AAACGGGCAAACTGGCACTTTAGG-1 | BL05111 |  8204 | 3595 |  12555 | 121 | BL | AAACGGGCAAACTGGCACTTTAGG-1 | 0.02644231 | False | 0.02644231 | False | 0.02644231 | False |\n",
       "| AAACGGGCAGGATCCGACTTTAGG-1 | BL05111 |   840 |  674 |  70401 | 130 | BL | AAACGGGCAGGATCCGACTTTAGG-1 | 0.02151755 | False | 0.02151755 | False | 0.02151755 | False |\n",
       "\n"
      ],
      "text/plain": [
       "                           orig.ident nCount_RNA nFeature_RNA nCount_ADT\n",
       "AAACAAGCAATTGAACACTTTAGG-1 BL05111    13302      4388          13535    \n",
       "AAACAAGCACAAATACACTTTAGG-1 BL05111     6375      3124          10241    \n",
       "AAACAAGCATACTTCAACTTTAGG-1 BL05111      625       431         144311    \n",
       "AAACCAATCACTATGTACTTTAGG-1 BL05111     6118      3041          17478    \n",
       "AAACCAATCATGAGGTACTTTAGG-1 BL05111    22733      5900          14822    \n",
       "AAACCAATCCGCTATAACTTTAGG-1 BL05111     9181      3722          12917    \n",
       "AAACCAATCTGTGACCACTTTAGG-1 BL05111     7964      3618          19663    \n",
       "AAACCAGGTGACATTCACTTTAGG-1 BL05111     8798      3849          10989    \n",
       "AAACGGGCAAACTGGCACTTTAGG-1 BL05111     8204      3595          12555    \n",
       "AAACGGGCAGGATCCGACTTTAGG-1 BL05111      840       674          70401    \n",
       "                           nFeature_ADT type Barcodes                  \n",
       "AAACAAGCAATTGAACACTTTAGG-1 127          BL   AAACAAGCAATTGAACACTTTAGG-1\n",
       "AAACAAGCACAAATACACTTTAGG-1 125          BL   AAACAAGCACAAATACACTTTAGG-1\n",
       "AAACAAGCATACTTCAACTTTAGG-1 130          BL   AAACAAGCATACTTCAACTTTAGG-1\n",
       "AAACCAATCACTATGTACTTTAGG-1 129          BL   AAACCAATCACTATGTACTTTAGG-1\n",
       "AAACCAATCATGAGGTACTTTAGG-1 128          BL   AAACCAATCATGAGGTACTTTAGG-1\n",
       "AAACCAATCCGCTATAACTTTAGG-1 124          BL   AAACCAATCCGCTATAACTTTAGG-1\n",
       "AAACCAATCTGTGACCACTTTAGG-1 128          BL   AAACCAATCTGTGACCACTTTAGG-1\n",
       "AAACCAGGTGACATTCACTTTAGG-1 125          BL   AAACCAGGTGACATTCACTTTAGG-1\n",
       "AAACGGGCAAACTGGCACTTTAGG-1 121          BL   AAACGGGCAAACTGGCACTTTAGG-1\n",
       "AAACGGGCAGGATCCGACTTTAGG-1 130          BL   AAACGGGCAGGATCCGACTTTAGG-1\n",
       "                           Doublet_Score Predicted_Doublet Doublet_Score.x\n",
       "AAACAAGCAATTGAACACTTTAGG-1 0.03723764    False             0.03723764     \n",
       "AAACAAGCACAAATACACTTTAGG-1 0.08790072    False             0.08790072     \n",
       "AAACAAGCATACTTCAACTTTAGG-1 0.01258178    False             0.01258178     \n",
       "AAACCAATCACTATGTACTTTAGG-1 0.09899889    False             0.09899889     \n",
       "AAACCAATCATGAGGTACTTTAGG-1 0.03723764    False             0.03723764     \n",
       "AAACCAATCCGCTATAACTTTAGG-1 0.09052632    False             0.09052632     \n",
       "AAACCAATCTGTGACCACTTTAGG-1 0.03614458    False             0.03614458     \n",
       "AAACCAGGTGACATTCACTTTAGG-1 0.03507611    False             0.03507611     \n",
       "AAACGGGCAAACTGGCACTTTAGG-1 0.02644231    False             0.02644231     \n",
       "AAACGGGCAGGATCCGACTTTAGG-1 0.02151755    False             0.02151755     \n",
       "                           Predicted_Doublet.x Doublet_Score.y\n",
       "AAACAAGCAATTGAACACTTTAGG-1 False               0.03723764     \n",
       "AAACAAGCACAAATACACTTTAGG-1 False               0.08790072     \n",
       "AAACAAGCATACTTCAACTTTAGG-1 False               0.01258178     \n",
       "AAACCAATCACTATGTACTTTAGG-1 False               0.09899889     \n",
       "AAACCAATCATGAGGTACTTTAGG-1 False               0.03723764     \n",
       "AAACCAATCCGCTATAACTTTAGG-1 False               0.09052632     \n",
       "AAACCAATCTGTGACCACTTTAGG-1 False               0.03614458     \n",
       "AAACCAGGTGACATTCACTTTAGG-1 False               0.03507611     \n",
       "AAACGGGCAAACTGGCACTTTAGG-1 False               0.02644231     \n",
       "AAACGGGCAGGATCCGACTTTAGG-1 False               0.02151755     \n",
       "                           Predicted_Doublet.y\n",
       "AAACAAGCAATTGAACACTTTAGG-1 False              \n",
       "AAACAAGCACAAATACACTTTAGG-1 False              \n",
       "AAACAAGCATACTTCAACTTTAGG-1 False              \n",
       "AAACCAATCACTATGTACTTTAGG-1 False              \n",
       "AAACCAATCATGAGGTACTTTAGG-1 False              \n",
       "AAACCAATCCGCTATAACTTTAGG-1 False              \n",
       "AAACCAATCTGTGACCACTTTAGG-1 False              \n",
       "AAACCAGGTGACATTCACTTTAGG-1 False              \n",
       "AAACGGGCAAACTGGCACTTTAGG-1 False              \n",
       "AAACGGGCAGGATCCGACTTTAGG-1 False              "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(fully)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e20e39d3-5d7b-4cb1-87c2-d9ac40b45022",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Figure 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d51207ce-d07f-4195-8541-9b047582ee03",
   "metadata": {},
   "outputs": [],
   "source": [
    "fully$sample2 <- fully$orig.ident\n",
    "\n",
    "fully$sample2 <- gsub('BL05111','CWB-1', fully$sample2)\n",
    "fully$sample2 <- gsub('BL05113','CWB-2', fully$sample2)\n",
    "fully$sample2 <- gsub('BL05401','drop', fully$sample2) # dropping Sample05401\n",
    "fully$sample2 <- gsub('BL05402','CWB-3', fully$sample2)\n",
    "fully$sample2 <- gsub('PB05111','Ficoll-1', fully$sample2)\n",
    "fully$sample2 <- gsub('PB05113','Ficoll-2', fully$sample2)\n",
    "fully$sample2 <- gsub('PB05401','drop', fully$sample2) # dropping Sample05401\n",
    "fully$sample2 <- gsub('PB05402','Ficoll-3', fully$sample2)\n",
    "\n",
    "fully <- subset(fully, subset=sample2!='drop')\n",
    "\n",
    "fully$pres <- fully$type\n",
    "\n",
    "fully$pres <- gsub('BL','CWB',fully$pres)\n",
    "fully$pres <- gsub('PB','Ficoll',fully$pres)\n",
    "\n",
    "fully$sample3 <- fully$orig.ident\n",
    "\n",
    "fully$sample3 <- gsub('BL05111','Sample-1', fully$sample3)\n",
    "fully$sample3 <- gsub('BL05113','Sample-2', fully$sample3)\n",
    "\n",
    "fully$sample3 <- gsub('BL05402','Sample-3', fully$sample3)\n",
    "fully$sample3 <- gsub('PB05111','Sample-1', fully$sample3)\n",
    "fully$sample3 <- gsub('PB05113','Sample-2', fully$sample3)\n",
    "\n",
    "fully$sample3 <- gsub('PB05402','Sample-3', fully$sample3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "49ae3ab8-a83d-4314-a849-5f6eacea8409",
   "metadata": {},
   "outputs": [],
   "source": [
    "yl_or <- '#FAC05E'\n",
    "blu <- '#1593CD'\n",
    "sage <- '#91C7B1'\n",
    "red <- '#B33951'\n",
    "pal4 <- c(yl_or,blu,sage,red)\n",
    "pal2 <- c(blu,red)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8a3b98b4-aae2-4379-8d64-74fc79c14787",
   "metadata": {},
   "outputs": [],
   "source": [
    "fully$pres <- factor(fully$pres, levels = c('Ficoll','CWB'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ad4df839-f269-4cbd-9204-210b0e17cf5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fully[[\"percent.mt\"]] <- PercentageFeatureSet(fully, pattern = \"^MT-\")\n",
    "fully <- subset(fully, subset=percent.mt < 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f2ca0db0-947b-467f-b84d-34ccf582c99a",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Normalizing layer: counts\n",
      "\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  NCF2, SPI1, SERPINA1, CST3, IFI30, CLEC7A, MNDA, EMILIN2, HCK, CD68 \n",
      "\t   PLXNB2, LILRB2, TNFAIP2, LRRC25, ZNF385A, TYMP, MEFV, SULF2, FGL2, MPEG1 \n",
      "\t   CYBB, CPVL, LYZ, TIMP2, CSTA, CSF3R, PILRA, LRP1, SLC7A7, KLF4 \n",
      "Negative:  IL7R, CD247, LTB, LEF1, CD7, TRBC1, PYHIN1, TRAT1, SAMD3, CD28 \n",
      "\t   PCED1B, CD69, CD27, CCR7, TRIB2, BCL2, ITGA6, MAL, IL2RB, TRABD2A \n",
      "\t   KLRK1, CTSW, OBSCN, ITM2A, HIST1H1C, MYC, NELL2, KLRG1, PRF1, CST7 \n",
      "PC_ 2 \n",
      "Positive:  NIBAN3, BANK1, CD79A, IGHM, FCRL1, MS4A1, CD22, PAX5, IGHD, FCRL2 \n",
      "\t   TCL1A, TNFRSF13C, BCL11A, RUBCNL, POU2AF1, RALGPS2, FCER2, BLK, COBLL1, FCRLA \n",
      "\t   AFF3, CD79B, BLNK, CXCR5, TSPAN13, IGKC, OSBPL10, WDFY4, COL19A1, ADAM28 \n",
      "Negative:  NKG7, PRF1, CST7, KLRD1, GZMA, SRGN, CCL5, CTSW, SYNE1, ADGRG1 \n",
      "\t   SAMD3, S100A4, MYO1F, TBX21, IL2RB, GNLY, FGFBP2, ANXA1, TGFBR3, CX3CR1 \n",
      "\t   GZMH, C1orf21, MYBL1, KLRK1, FCRL6, PYHIN1, PRSS23, CD300A, PLEKHG3, MATK \n",
      "PC_ 3 \n",
      "Positive:  NKG7, GZMB, KLRD1, ADGRG1, PRF1, PLEK, CST7, TBX21, GNLY, FGFBP2 \n",
      "\t   SYNE1, GZMA, C1orf21, ADRB2, ADGRG5, SLAMF7, PRSS23, ZEB2, CX3CR1, GZMH \n",
      "\t   FCRL6, CCL5, MCTP2, MAP3K8, HOPX, CTSW, S1PR5, CLIC3, MATK, TGFBR3 \n",
      "Negative:  MAL, LEF1, IL7R, TRABD2A, CCR7, CD28, ACTN1, CD4, NELL2, LTB \n",
      "\t   ITGA6, MYC, TRAT1, CD27, IL6R, RNF157, RGCC, NOG, FHIT, TCEA3 \n",
      "\t   CD40LG, OBSCN, ADTRP, AQP3, VIM, ICOS, LRRN3, HAPLN3, SESN3, PCED1B \n",
      "PC_ 4 \n",
      "Positive:  CLEC4C, PTPRS, SERPINF1, TNFRSF21, LILRA4, SCAMP5, EPHA2, PPM1J, SCN9A, LAMP5 \n",
      "\t   SCT, IL3RA, DNASE1L3, NRP1, GAS6, PACSIN1, EPHB1, P3H2, MAP1A, PTCRA \n",
      "\t   PAPLN, LRRC26, TIFAB, PHEX, SLC41A2, FAM160A1, SLC12A3, SMPD3, PLD4, KCNK17 \n",
      "Negative:  FCRL1, MS4A1, PAX5, FCRL2, CD22, FCER2, IGHD, CD79A, CD79B, TNFRSF13C \n",
      "\t   POU2AF1, COL19A1, BANK1, SWAP70, CXCR5, OSBPL10, SNX22, CD19, FCRLA, CD40 \n",
      "\t   P2RX5, RALGPS2, MTSS1, EBF1, CD24, SEMA4B, ADAM28, CLEC17A, STAP1, CD72 \n",
      "PC_ 5 \n",
      "Positive:  S100A12, QPCT, PADI4, CRISPLD2, PADI2, CLEC4E, CD14, CLEC4D, CD163, S100A8 \n",
      "\t   VCAN, MCEMP1, MARC1, MGST1, RNASE2, DYSF, FOLR3, CES1, CYP1B1, RNASE6 \n",
      "\t   ASGR2, TREM1, WLS, PLA2G7, FCAR, CCR2, SIRPA, STEAP4, ITGAM, STAB1 \n",
      "Negative:  CDKN1C, TCF7L2, HES4, CKB, NEURL1, SIGLEC10, ZNF703, PTP4A3, FCGR3A, CTSL \n",
      "\t   FMNL2, FCGR3B, CEACAM3, TAGLN, MYOF, IFITM3, KNDC1, PAPSS2, ADGRE1, LYPD2 \n",
      "\t   MS4A4A, MS4A7, MERTK, CLEC4F, TBC1D8, MTSS1, MGLL, RHOC, TNFRSF8, LPL \n",
      "\n",
      "Warning message:\n",
      "“The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
      "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
      "This message will be shown once per session”\n",
      "17:56:36 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "17:56:36 Read 42846 rows and found 20 numeric columns\n",
      "\n",
      "17:56:36 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "17:56:36 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "17:56:41 Writing NN index file to temp file /tmp/RtmpWpIAwe/file10d905792c4eb\n",
      "\n",
      "17:56:41 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "17:56:57 Annoy recall = 100%\n",
      "\n",
      "17:56:57 Commencing smooth kNN distance calibration using 1 thread\n",
      "\n",
      "17:56:58 Initializing from normalized Laplacian + noise\n",
      "\n",
      "17:57:03 Commencing optimization for 200 epochs, with 1849656 positive edges\n",
      "\n",
      "17:57:22 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 42846\n",
      "Number of edges: 1458900\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9335\n",
      "Number of communities: 21\n",
      "Elapsed time: 17 seconds\n"
     ]
    }
   ],
   "source": [
    "fully <- NormalizeData(fully) %>% \n",
    "    FindVariableFeatures() %>% \n",
    "    ScaleData() %>% \n",
    "    RunPCA() %>% \n",
    "    RunUMAP(dims = 1:20) %>% \n",
    "    FindNeighbors(dims = 1:20) %>% \n",
    "    FindClusters(resolution = 0.5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "774e24ba-1505-4dc7-9cd3-0e8160d87aff",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Celltype Label Transfer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8dce7208-b675-4411-8f9d-1ab9527b7a44",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Validating h5Seurat file\n",
      "\n",
      "Initializing ADT with data\n",
      "\n",
      "Adding counts for ADT\n",
      "\n",
      "Adding variable feature information for ADT\n",
      "\n",
      "Adding miscellaneous information for ADT\n",
      "\n",
      "Initializing SCT with data\n",
      "\n",
      "Adding counts for SCT\n",
      "\n",
      "Adding variable feature information for SCT\n",
      "\n",
      "Adding miscellaneous information for SCT\n",
      "\n",
      "Adding reduction apca\n",
      "\n",
      "Adding cell embeddings for apca\n",
      "\n",
      "Adding feature loadings for apca\n",
      "\n",
      "Adding miscellaneous information for apca\n",
      "\n",
      "Adding reduction aumap\n",
      "\n",
      "Adding cell embeddings for aumap\n",
      "\n",
      "Adding miscellaneous information for aumap\n",
      "\n",
      "Adding reduction pca\n",
      "\n",
      "Adding cell embeddings for pca\n",
      "\n",
      "Adding feature loadings for pca\n",
      "\n",
      "Adding miscellaneous information for pca\n",
      "\n",
      "Adding reduction spca\n",
      "\n",
      "Adding cell embeddings for spca\n",
      "\n",
      "Adding feature loadings for spca\n",
      "\n",
      "Adding miscellaneous information for spca\n",
      "\n",
      "Adding reduction umap\n",
      "\n",
      "Adding cell embeddings for umap\n",
      "\n",
      "Adding miscellaneous information for umap\n",
      "\n",
      "Adding reduction wnn.umap\n",
      "\n",
      "Adding cell embeddings for wnn.umap\n",
      "\n",
      "Adding miscellaneous information for wnn.umap\n",
      "\n",
      "Adding graph wknn\n",
      "\n",
      "Adding graph wsnn\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ref <- LoadH5Seurat(file = '/home/jupyter/pbmc_multimodal.h5seurat')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3475f651-3c82-4ce2-9a4e-39de526625bc",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Normalizing query using reference SCT model\n",
      "\n",
      "Projecting cell embeddings\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 6230 anchors\n",
      "\n"
     ]
    }
   ],
   "source": [
    "anchors <- FindTransferAnchors(\n",
    "    reference = ref,\n",
    "    query = fully,\n",
    "    normalization.method = \"SCT\",\n",
    "    reference.reduction = \"spca\",\n",
    "    dims = 1:50\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6a931087-4f2e-43ae-8cc2-f3c6d68a68fb",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Predicting cell labels\n",
      "\n",
      "Warning message:\n",
      "“Layer counts isn't present in the assay object; returning NULL”\n",
      "Predicting cell labels\n",
      "\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Layer counts isn't present in the assay object; returning NULL”\n",
      "Predicting cell labels\n",
      "\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Warning message:\n",
      "“Layer counts isn't present in the assay object; returning NULL”\n",
      "Transfering 228 features onto reference data\n",
      "\n",
      "Warning message:\n",
      "“Layer counts isn't present in the assay object; returning NULL”\n"
     ]
    }
   ],
   "source": [
    "fully <- TransferData(\n",
    "    anchorset = anchors, \n",
    "    reference = ref, \n",
    "    query = fully,\n",
    "    refdata = list(\n",
    "        celltype.l1 = \"celltype.l1\",\n",
    "        celltype.l2 = \"celltype.l2\",\n",
    "        celltype.l3 = \"celltype.l3\",\n",
    "        predicted_ADT = 'ADT'\n",
    "    )\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "54c73f1d-6a2f-42fa-8f1e-4b14c605dbd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(fully)<- 'ADT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "114de83a-4188-41ab-abb5-cbfe9169b862",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "adt_cluster <- function(so){\n",
    "    DefaultAssay(so) <- 'ADT'\n",
    "    so <- so %>%\n",
    "    NormalizeData(normalization.method='CLR',margin=2) %>%\n",
    "    ScaleData() %>%\n",
    "    RunPCA(features = rownames(so), reduction.name = \"pca_adt\", reduction.key = \"pca_adt_\", verbose = FALSE) %>% \n",
    "    RunUMAP(dims = 1:20, reduction='pca_adt', assay='ADT', reduction.name = 'adt.UMAP', reduction.key='adtUMAP_') %>% \n",
    "    FindNeighbors(dims = 1:20, graph.name = 'adt_snn') %>% \n",
    "    FindClusters(resolution = 0.5, 'adt_snn')\n",
    "    return(so)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9ace2a41-6bc1-4f5a-a765-b2507e828990",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Normalizing across cells\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message:\n",
      "“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pca_adt_ to pcaadt_”\n",
      "18:04:36 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "18:04:36 Read 42846 rows and found 20 numeric columns\n",
      "\n",
      "18:04:36 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "18:04:36 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "18:04:40 Writing NN index file to temp file /tmp/RtmpWpIAwe/file10d90629e609b\n",
      "\n",
      "18:04:40 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "18:04:56 Annoy recall = 100%\n",
      "\n",
      "18:04:58 Commencing smooth kNN distance calibration using 1 thread\n",
      "\n",
      "18:04:59 Initializing from normalized Laplacian + noise\n",
      "\n",
      "18:05:03 Commencing optimization for 200 epochs, with 1869158 positive edges\n",
      "\n",
      "18:05:22 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n",
      "Only one graph name supplied, storing nearest-neighbor graph only\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 42846\n",
      "Number of edges: 400334\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9179\n",
      "Number of communities: 43\n",
      "Elapsed time: 7 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25 singletons identified. 18 final clusters.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "fully <- adt_cluster(fully)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3badb74b-cfd2-4052-a283-269c1954e1a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(fully,'/home/jupyter/CS15_WHBL/CWB_Paper/01_Final_Data/01_Data/Fig_1_Final.rds') # writing \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1e9a6b32-ee66-4d8e-b633-85982847bb73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.1.3 (2022-03-10)\n",
       "Platform: x86_64-pc-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.4 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3\n",
       "LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       \n",
       " [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   \n",
       " [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          \n",
       "[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   \n",
       "\n",
       "attached base packages:\n",
       "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] ggrepel_0.9.5         SeuratDisk_0.0.0.9020 lubridate_1.9.3      \n",
       " [4] forcats_1.0.0         stringr_1.5.1         purrr_1.0.2          \n",
       " [7] readr_2.1.5           tidyr_1.3.1           tibble_3.2.1         \n",
       "[10] ggplot2_3.4.3         tidyverse_2.0.0       hise_2.15.0          \n",
       "[13] H5weaver_1.2.0        rhdf5_2.38.1          Matrix_1.6-4         \n",
       "[16] data.table_1.14.2     patchwork_1.1.1       Seurat_5.1.0         \n",
       "[19] SeuratObject_5.0.2    sp_2.1-4              dplyr_1.1.4          \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] Rtsne_0.16             colorspace_2.0-3       deldir_1.0-6          \n",
       "  [4] ellipsis_0.3.2         ggridges_0.5.3         IRdisplay_1.1         \n",
       "  [7] RcppHNSW_0.3.0         base64enc_0.1-3        spatstat.data_3.0-4   \n",
       " [10] leiden_0.3.10          listenv_0.8.0          bit64_4.0.5           \n",
       " [13] RSpectra_0.16-1        fansi_1.0.3            codetools_0.2-18      \n",
       " [16] splines_4.1.3          polyclip_1.10-0        IRkernel_1.3          \n",
       " [19] spam_2.8-0             jsonlite_1.8.8         ica_1.0-2             \n",
       " [22] cluster_2.1.3          png_0.1-7              uwot_0.1.11           \n",
       " [25] shiny_1.7.1            sctransform_0.4.1      spatstat.sparse_3.0-3 \n",
       " [28] compiler_4.1.3         httr_1.4.7             fastmap_1.2.0         \n",
       " [31] lazyeval_0.2.2         cli_3.6.2              later_1.3.0           \n",
       " [34] htmltools_0.5.2        tools_4.1.3            igraph_1.3.1          \n",
       " [37] dotCall64_1.0-1        gtable_0.3.0           glue_1.6.2            \n",
       " [40] RANN_2.6.1             reshape2_1.4.4         Rcpp_1.0.8.3          \n",
       " [43] scattermore_1.2        rhdf5filters_1.6.0     vctrs_0.6.5           \n",
       " [46] nlme_3.1-157           spatstat.explore_3.2-7 progressr_0.10.0      \n",
       " [49] lmtest_0.9-40          spatstat.random_3.2-3  globals_0.14.0        \n",
       " [52] timechange_0.3.0       mime_0.12              miniUI_0.1.1.1        \n",
       " [55] lifecycle_1.0.4        irlba_2.3.5            goftest_1.2-3         \n",
       " [58] future_1.25.0          MASS_7.3-57            zoo_1.8-10            \n",
       " [61] scales_1.3.0           hms_1.1.3              promises_1.2.0.1      \n",
       " [64] spatstat.utils_3.0-4   parallel_4.1.3         RColorBrewer_1.1-3    \n",
       " [67] reticulate_1.24        pbapply_1.5-0          gridExtra_2.3         \n",
       " [70] stringi_1.7.6          fastDummies_1.7.3      repr_1.1.4.9000       \n",
       " [73] rlang_1.1.3            pkgconfig_2.0.3        matrixStats_0.62.0    \n",
       " [76] evaluate_0.15          lattice_0.20-45        Rhdf5lib_1.16.0       \n",
       " [79] ROCR_1.0-11            tensor_1.5             htmlwidgets_1.5.4     \n",
       " [82] bit_4.0.4              cowplot_1.1.1          tidyselect_1.2.1      \n",
       " [85] parallelly_1.31.1      RcppAnnoy_0.0.19       plyr_1.8.7            \n",
       " [88] magrittr_2.0.3         R6_2.5.1               generics_0.1.3        \n",
       " [91] pbdZMQ_0.3-7           withr_2.5.0            pillar_1.9.0          \n",
       " [94] fitdistrplus_1.1-8     survival_3.3-1         abind_1.4-5           \n",
       " [97] future.apply_1.9.0     hdf5r_1.3.5            crayon_1.5.1          \n",
       "[100] uuid_1.1-0             KernSmooth_2.23-20     utf8_1.2.2            \n",
       "[103] spatstat.geom_3.2-9    plotly_4.10.0          tzdb_0.4.0            \n",
       "[106] grid_4.1.3             digest_0.6.29          xtable_1.8-4          \n",
       "[109] httpuv_1.6.5           munsell_0.5.0          viridisLite_0.4.0     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "474d4106-da5f-43d7-9525-4bc0e8edf60b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
